File: client/py/synnax/cli/io.py
#  Copyright 2023 Synnax Labs, Inc.
#
#  Use of this software is governed by the Business Source License included in the file
#  licenses/BSL.txt.
#
#  As of the Change Date specified in that file, in accordance with the Business Source
#  License, use of this software will be governed by the Apache License, Version 2.0,
#  included in the file licenses/APL.txt.

from pathlib import Path

from synnax.cli.flow import Context
from synnax.io import IO_FACTORY, RowFileReader


def prompt_file(ctx: Context) -> Path | None:
    """Prompts the user for a file path."""
    _fp = ctx.console.ask("File path")
    assert _fp is not None
    fp = Path(_fp)
    if not fp.exists():
        ctx.console.error(f"File does not exist: {fp}")
        return prompt_file(ctx)
    if not fp.is_file():
        ctx.console.error(f"Not a file: {fp}")
        return prompt_file(ctx)
    ctx.console.success(f"File found: {fp}")
    return fp


def prompt_new_reader(
    ctx: Context, path: Path | str | None = None
) -> RowFileReader | None:
    """Prompts the user for a file path and returns a new reader for that file."""
    if path is None:
        path = prompt_file(ctx)
    if path is None:
        return None
    return IO_FACTORY.new_reader(path if isinstance(path, Path) else Path(path))

File: x/go/io/io.go
// Copyright 2025 Synnax Labs, Inc.
//
// Use of this software is governed by the Business Source License included in the file
// licenses/BSL.txt.
//
// As of the Change Date specified in that file, in accordance with the Business Source
// License, use of this software will be governed by the Apache License, Version 2.0,
// included in the file licenses/APL.txt.

package io

import "io"

type ReaderAtCloser interface {
	io.ReaderAt
	io.Closer
}

type WriterAtCloser interface {
	io.WriterAt
	io.Closer
}

type ReaderAtWriterAtCloser interface {
	io.ReaderAt
	WriterAtCloser
}

type TrackedWriteCloser interface {
	OffsetWriter
	io.Closer
}

type OffsetWriter interface {
	// Reset resets the offset of the writer to the current offset.
	Reset()
	// Offset returns the offset of the write cursor.
	Offset() int64
	// Len returns the number of bytes written by the writer.
	Len() int64
	io.Writer
}

type combinedReadWriteCloser struct {
	io.Reader
	io.Writer
	io.Closer
}

func (c combinedReadWriteCloser) Close() error {
	if c.Closer != nil {
		return c.Closer.Close()
	}
	return nil
}

func CombineReadWriteCloser(r io.Reader, w io.Writer, closer io.Closer) io.ReadWriteCloser {
	return combinedReadWriteCloser{Reader: r, Writer: w, Closer: closer}
}
